src/util/include/util/DispatchPool.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Xournal++
+ *
+ * Template class for a dispatch pool and listeners
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <vector>
+
+namespace xoj::util {
+
+/**
+ * @brief Generic template for dispatchers.
+ *
+ * The listeners must implement functions
+ *      void on(Arg&&...);
+ * that the dispatcher will call through dispatch().
+ */
+template <class ListenerT>
+class DispatchPool final {
+public:
+    using listener_type = ListenerT;
+
+    template <typename... Args>
+    void dispatch(Args&&... args) const {
+        for (auto* v: this->pool) {
+            v->on(std::forward<Args>(args)...);
+        }
+    }
+
+    void add(listener_type* v) {
+        assert(v != nullptr && "Adding nullptr listener");
+        assert(std::find(this->pool.begin(), this->pool.end(), v) == this->pool.end() && "Listener is already listed");
+        this->pool.emplace_back(v);
+    }
+
+    void remove(listener_type* v) {
+        assert(v != nullptr && "Removing nullptr listener");
+        auto it = std::find(this->pool.begin(), this->pool.end(), v);
+        if (it != this->pool.end()) {
+            this->pool.erase(it);
+        }
+    }
+
+    [[nodiscard]] bool empty() const { return pool.empty(); }
+
+private:
+    std::vector<listener_type*> pool;
+};
+
+/**
+ * @brief CRTP-style class for listener
+ * Usage:
+ *  class A: Listener<A> {
+ *      (virtual) ~A() { unregisterFromPool(); }
+ *      void on(...);  // Signal receiver
+ *  };
+ *
+ * WARNING: Always unregisterFromPool() in derived class destructor.
```

Comment:
This seems like a pretty big caveat. Would it be a big problem to add a virtual destructor in Listener that automatically does this?

Reply:
I'd rather not impose a vtable on every listener type.
More importantly, this would open a data race upon destruction:
the base class gets destroyed last, so in a multithreading situation, a function on() of the derived class could be called (by the dispatcher) between the destruction of the derived class and before the destruction of the base (hence the call to unregisterFromPool()).

I don't have a proper solution for this.